{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "### 1. FindORFs.py"
      ],
      "metadata": {
        "id": "OqUwG7ab_Hsb"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "6ULaXNGN-35x"
      },
      "outputs": [],
      "source": [
        "class CommandLine():\n",
        "    \"\"\"\n",
        "    Handle the command line, usage and help requests.\n",
        "\n",
        "    CommandLine uses argparse, now standard in 2.7 and beyond.\n",
        "    it implements a standard command line argument parser with various argument options,\n",
        "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
        "\n",
        "    attributes:\n",
        "    all arguments received from the commandline using .add_argument will be\n",
        "    avalable within the .args attribute of object instantiated from CommandLine.\n",
        "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
        "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
        "    name that option.\n",
        "    \"\"\"\n",
        "    def __init__(self, inOpts=None):\n",
        "        \"\"\"\n",
        "         CommandLine constructor.\n",
        "        Implements a parser to interpret the command line argv string using argparse.\n",
        "        :param inOpts:\n",
        "        \"\"\"\n",
        "        import argparse\n",
        "        self.parser = argparse.ArgumentParser(\n",
        "            description='Finds the largest ORF in a DNA sequence',\n",
        "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
        "            add_help=True,  # default is True\n",
        "            prefix_chars='-',\n",
        "            usage='%(prog)s [options] -option1[default] <input >output'\n",
        "            )\n",
        "        self.parser.add_argument('-lG', '--longestGene', action='store_true', help='longest Gene in an ORF')\n",
        "        self.parser.add_argument('-mG', '--minGene', type=int, choices=(100, 200, 300, 500, 1000), action='store',\n",
        "                                 help='minimum Gene length')\n",
        "        self.parser.add_argument('-s', '--start', action='append', nargs='?',\n",
        "                                 help='start Codon')  # allows multiple list options\n",
        "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
        "        if inOpts is None:\n",
        "            self.args = self.parser.parse_args()\n",
        "        else:\n",
        "            self.args = self.parser.parse_args(inOpts)\n",
        "\n",
        "class OrfFinder():\n",
        "    \"\"\"\n",
        "    Takes in a sequence of a FASTA file and stores ORFs in a lists of lists.\n",
        "    Attributes:\n",
        "        attr1 (list): List of valid stop codons.\n",
        "        attr2 (list): List of valid start codon.\n",
        "        attr3 (dict): Dictionary of valid nucleotides.\n",
        "    \"\"\"\n",
        "    stop_codons = ['TGA', 'TAG', 'TAA']\n",
        "    start_codons = ['ATG']\n",
        "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
        "\n",
        "    def __init__(self, seq):\n",
        "        \"\"\"Takes in sequence from users input file then creates a list of\n",
        "        lists.\n",
        "        Then creates ORFs list  which holds the frame, start, stop, length of the ORF.\n",
        "        Args:\n",
        "            param1 (str): String of DNA from FASTA file.\n",
        "        \"\"\"\n",
        "        self.seq = seq\n",
        "        self.orfs = []  # The lists where we save lists of ORFS and their position/length/frame.\n",
        "\n",
        "    def findOrfs(self):\n",
        "        \"\"\"\n",
        "        Find Orfs on top strand and return list of Orfs.\n",
        "        Returns:\n",
        "            (list): list of lists of ORFs found.\n",
        "        \"\"\"\n",
        "        start_positions = []\n",
        "        foundStart = 0\n",
        "        foundCodon = 0\n",
        "\n",
        "        for frame in range(0,3):  # We need to check for frames 1, 2, 3\n",
        "            foundStart = 0  # Flag name for when finding codons and start codons.\n",
        "            foundCodon = 0\n",
        "            start_positions = []  # Clears the start position list for each frame.\n",
        "            for i in range(frame, len(self.seq), 3):\n",
        "                codon = self.seq[i:i+3]  # The codon is 3 nucleotides.\n",
        "                if codon == 'ATG':  # When start codon is found.\n",
        "                    start_positions.append(i)\n",
        "                    foundStart = 1\n",
        "                    foundCodon = 1\n",
        "\n",
        "                if codon in OrfFinder.stop_codons and foundStart:\n",
        "                    #print('ORF found')\n",
        "                    start = start_positions[0] + 1 - frame\n",
        "                    stop = i + 3\n",
        "                    length = stop - start + 1\n",
        "                    self.saveOrf((frame%3) + 1, start, stop, length)\n",
        "                    start_positions = []\n",
        "                    foundStart = 0\n",
        "                    foundCodon = 1\n",
        "\n",
        "                if not foundCodon and codon in OrfFinder.stop_codons:  # If no start codon was found but stop codon found.\n",
        "                    #print(\"Dangling stop at 5' end\")\n",
        "                    start = 1\n",
        "                    stop = i + 3\n",
        "                    length = stop - start + 1\n",
        "                    self.saveOrf((frame%3) + 1, start, stop, length)\n",
        "                    start_positions = []\n",
        "                    foundCodon = 1\n",
        "\n",
        "            if foundStart:  # If no stop codon was found but start codon was found.\n",
        "                #print(\"Dangling start at 3' end\")\n",
        "                start = start_positions[0] + 1\n",
        "                stop = len(self.seq)\n",
        "                length = stop - start + 1\n",
        "                self.saveOrf((frame%3) + 1, start, stop, length)\n",
        "\n",
        "        return self.orfs\n",
        "\n",
        "    def findRevOrfs(self):\n",
        "        \"\"\" Find Orfs on the bottom strand and returns that list of Orfs.\n",
        "        Remember to fixup the orfList so that it refers to top strand coordinates and the rev frames.\n",
        "        Returns:\n",
        "            (list): list of lists of ORFs found on the reverse complement.\n",
        "        \"\"\"\n",
        "        reverseComp = self.reverseComplement()\n",
        "        start_positions = []\n",
        "        foundStart = 0\n",
        "        foundCodon = 0\n",
        "\n",
        "        for frame in range(0, 3):  # We need to check for frames 1, 2, 3\n",
        "            foundStart = 0  # Flag name for when finding codons and start codons.\n",
        "            foundCodon = 0\n",
        "            start_positions = []  # Clears the start position list for each frame.\n",
        "            for i in range(frame, len(reverseComp), 3):\n",
        "                codon = reverseComp[i:i + 3]  # The codon is 3 nucleotides.\n",
        "\n",
        "                if codon == 'ATG':  # When start codon is found.\n",
        "                    start_positions.append(i)\n",
        "                    foundStart = 1\n",
        "                    foundCodon = 1\n",
        "\n",
        "                if codon in OrfFinder.stop_codons and foundStart:  # We have a start and stop codon.\n",
        "                    stop = len(reverseComp) - start_positions[0]\n",
        "                    start = len(reverseComp) - (i+2)\n",
        "                    if frame == 1: stop += 1  # I'm not sure about this part.\n",
        "                    elif frame == 2: stop += 2\n",
        "                    length = stop - start + 1\n",
        "                    self.saveOrf(-1 * ((frame%3) + 1), start, stop, length)\n",
        "                    start_positions = []\n",
        "                    foundStart = 0\n",
        "                    foundCodon = 1\n",
        "\n",
        "                if not foundCodon and codon in OrfFinder.stop_codons:  # If no start codon was found but stop codon found.\n",
        "                    #print(\"Dangling stop at 5' end\")\n",
        "                    start = len(reverseComp) - i - 2\n",
        "                    stop = len(reverseComp)\n",
        "                    length = stop - start + 1\n",
        "                    self.saveOrf(-1 * ((frame%3) + 1), start, stop, length)\n",
        "                    start_positions = []\n",
        "                    foundCodon = 1\n",
        "\n",
        "            if foundStart:  # If no stop codon was found but start codon was found.\n",
        "                #print(\"Dangling start at 3' end\")\n",
        "                start =  start_positions[0] + 1\n",
        "                stop = 1\n",
        "                length = stop - start + 1\n",
        "                self.saveOrf(-1 * ((frame%3) + 1), start, stop, length)\n",
        "\n",
        "        return self.orfs\n",
        "\n",
        "    def saveOrf(self, frame, start, stop, length):\n",
        "        \"\"\" Saves the information about the ORF that was found.\n",
        "        This method is used in findOrfs and findRevOrfs.\n",
        "        Args:\n",
        "            param1 (int): Start position.\n",
        "            param2 (int): Stop position.\n",
        "            param3 (int): Length of my Orf. (Stop - Start)\n",
        "            param4 (int): Which frame my Orf was found in.\n",
        "        Returns:\n",
        "            (list): A list of lists of Orfs with their appropriate information.\n",
        "        \"\"\"\n",
        "        self.orfs.append([frame, start, stop, length])\n",
        "\n",
        "    def reverseComplement(self):\n",
        "        \"\"\" Helper method to take the reverse complement of a DNA seqeunce.\n",
        "        Returns:\n",
        "            (list): Reverse complement of DNA sequence.\n",
        "        \"\"\"\n",
        "        return ''.join([self.complement[base] for base in self.seq[::-1]])  # Dictionary comprehension to find reverse complement of DNA sequence."
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 2. SequenceAnalysis.py"
      ],
      "metadata": {
        "id": "ch7--mnM_Pa2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "class NucParams:\n",
        "    \"\"\"Creates dictionaries to store fasta file information.\n",
        "\n",
        "        Attributes:\n",
        "            attr1 (dict): Dictionary of RNA codons.\n",
        "            attr2 (dict): Dictionary of DNA codons.\n",
        "            attr3 (dict): Dictionary of valid Nucleotides.\n",
        "    \"\"\"\n",
        "\n",
        "    rnaCodonTable = {\n",
        "                    # RNA codon table\n",
        "    # U\n",
        "        'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
        "        'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
        "        'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
        "        'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
        "    # C\n",
        "        'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
        "        'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
        "        'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
        "        'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
        "    # A\n",
        "        'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
        "        'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
        "        'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
        "        'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
        "    # G\n",
        "        'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
        "        'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
        "        'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
        "        'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'   # GxG\n",
        "    }\n",
        "\n",
        "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
        "\n",
        "    validNucleotides = {'A': 0, 'C': 0, 'G': 0, 'T': 0, 'U': 0, 'N':0}\n",
        "\n",
        "    def __init__(self):\n",
        "        \"\"\"\n",
        "        Initializes dictionaries with appropriate keys, and values set to zero.\n",
        "        \"\"\"\n",
        "        self.aminoAcidComposition = {}\n",
        "        self.codonsComposition = {}\n",
        "        self.nucleotideComposition = {}\n",
        "\n",
        "        for aa in ProteinParam.aa2mw:  # Initializes Amino Acid dictionary.\n",
        "            self.aminoAcidComposition[aa] = 0\n",
        "\n",
        "        for codon in self.rnaCodonTable:  # Initializes Codon Composition dictionary.\n",
        "            self.codonsComposition[codon] = 0\n",
        "\n",
        "        for nuc in NucParams.validNucleotides:  # Initializes Nucleotide Composition dictionary.\n",
        "            self.nucleotideComposition[nuc] = 0\n",
        "\n",
        "    def addSequence(self, thisSequence):\n",
        "        \"\"\"\n",
        "        This method accepts additional sequences and adds the data that you were given\n",
        "        with the __init__ method.\n",
        "        \"\"\"\n",
        "        for nuc in thisSequence:\n",
        "            if nuc in NucParams.validNucleotides.keys():  # Checks if each nucleotide is valid.\n",
        "                self.nucleotideComposition[nuc] += 1  # Counts how many nucleotides there are.\n",
        "\n",
        "        rnaSequence = thisSequence.replace('T', 'U')  # Converts DNA sequence to RNA\n",
        "\n",
        "        for start in range(0, len(rnaSequence), 3):\n",
        "            codon = rnaSequence[start: start + 3]\n",
        "            if codon in self.rnaCodonTable:\n",
        "                self.codonsComposition[codon] += 1  # Adds RNA sequence to dictionary w/ count.\n",
        "                aa = self.rnaCodonTable[codon]\n",
        "                if aa != '-':\n",
        "                    self.aminoAcidComposition[aa] += 1  # Adds amino acid to dictionary w/ count.\n",
        "\n",
        "\n",
        "    def aaComposition(self):\n",
        "        \"\"\"\n",
        "        This method will return a dictionary of counts over the 20 amino acids.\n",
        "        \"\"\"\n",
        "        return self.aminoAcidComposition\n",
        "\n",
        "    def nucComposition(self):\n",
        "        \"\"\"\n",
        "        Returns the updated nucleotide composition dictionary from addSequence.\n",
        "        \"\"\"\n",
        "        return self.nucleotideComposition\n",
        "\n",
        "    def codonComposition(self):\n",
        "        \"\"\"\n",
        "        This dictionary returns codons in RNA format and counts of codon.\n",
        "        \"\"\"\n",
        "        return self.codonsComposition\n",
        "\n",
        "    def nucCount(self):\n",
        "        \"\"\"\n",
        "        This returns an integer value, summing every valid nucleotide found. This\n",
        "        value should exactly equal the sum over the nucleotide composition dictionary.\n",
        "        \"\"\"\n",
        "        nucTotal = 0\n",
        "        for count in self.nucleotideComposition.values():\n",
        "            nucTotal += count\n",
        "        return nucTotal\n",
        "\n",
        "\n",
        "class ProteinParam(str):\n",
        "    \"\"\"Creates a ProteinParam string object in upper case letters.\n",
        "\n",
        "    Attributes:\n",
        "        attr1 (dict): Dictionary of molecular weights of amino acids.\n",
        "        attr2 (float): Molecular weight of H20.\n",
        "        attr3 (dict): Dictionary of absorbance values at 280nm.\n",
        "        attr4 (dict): Dictionary of positively charged amino acids.\n",
        "        attr5 (dict): Dictionary of negatively charged amino acids.\n",
        "        attr6 (float): Charge of the N terminus.\n",
        "        attr7 (float): Charge of the C terminus.\n",
        "        attr8 (dict): Dictionary of valid aa's.\n",
        "    \"\"\"\n",
        "\n",
        "    aa2mw = {\n",
        "        'A': 89.093, 'G': 75.067, 'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
        "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
        "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
        "        'W': 204.225, 'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
        "    }\n",
        "\n",
        "    mwH2O = 18.015\n",
        "    aa2abs280 = {'Y': 1490, 'W': 5500, 'C': 125}\n",
        "\n",
        "    aa2chargePos = {'K': 10.5, 'R': 12.4, 'H': 6}\n",
        "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
        "    aaNterm = 9.69\n",
        "    aaCterm = 2.34\n",
        "\n",
        "    def __init__(self, protein):\n",
        "        \"\"\"Takes in sequence from users input then creates a list of\n",
        "        strings and concatnates to create a string object.\n",
        "        Then creates aaDictionary with the keys being the valid aa's of\n",
        "        users input and the value of each one being their corresponding count.\n",
        "\n",
        "        Args:\n",
        "            param1 (str): String of amino acids.\n",
        "        \"\"\"\n",
        "        myList = ''.join(protein).split()\n",
        "        self.proteinString = ''.join(myList).upper()\n",
        "        self.aaDictionary = {}\n",
        "\n",
        "        for aa in self.aa2mw.keys():  # Iterates through aa2mw dictionary, the valid aa's.\n",
        "            self.aaDictionary[aa] = float(self.proteinString.count(aa))  # Stores values.\n",
        "\n",
        "    def aaCount(self):\n",
        "        \"\"\" Iterates through every character in string and returns a count of valid\n",
        "        amino acids. Ignores invalid characters.\n",
        "\n",
        "        Returns:\n",
        "            aaTotal (int): Total number of valid amino acids.\n",
        "        \"\"\"\n",
        "        aaTotal = 0\n",
        "        for aa in self.proteinString:\n",
        "            if aa.upper() in self.aa2mw.keys():  # Checks if character in string is a valid aa.\n",
        "                aaTotal += 1\n",
        "        return aaTotal\n",
        "\n",
        "    def pI(self):\n",
        "        \"\"\"Estimates the theoretical isoelectric point by finding the\n",
        "        particular pH that yeilds a neutral net charge (as close to\n",
        "        zero as possible, accurate to two decimal places).\n",
        "\n",
        "        Returns:\n",
        "            (float): Best pH.\n",
        "        \"\"\"\n",
        "        bigCharge = 2 ** 11  # Choose a big number, this is our loop invariant.\n",
        "        bestPH = 0\n",
        "        particularPH = 0\n",
        "        while particularPH < 14.01:\n",
        "            charge = abs(self.__charge__(particularPH))  # 0 <= charge <= 14\n",
        "            if charge < bigCharge:\n",
        "                bigCharge = charge\n",
        "                bestPH = particularPH\n",
        "            particularPH += 0.01  # Needs to iterate through every pH in decimal places.\n",
        "        return bestPH\n",
        "\n",
        "    def aaComposition(self):\n",
        "        \"\"\"\n",
        "        Returns:\n",
        "            aaDictionary created in __init__ method.\n",
        "        \"\"\"\n",
        "        return self.aaDictionary\n",
        "\n",
        "    def __charge__(self, pH):\n",
        "        \"\"\"Calculates the net charge on the protein at specific pH using pKa of each\n",
        "        charged amino acid, Nterminus and Cterminus.\n",
        "        Args:\n",
        "            param1 (float): pH value from pI method.\n",
        "        Returns:\n",
        "            (float): Net charge of the protein.\n",
        "        \"\"\"\n",
        "        posCharge = 0\n",
        "        for aa in self.aa2chargePos:  # Begin summation for posCharges.\n",
        "            nAA = self.aaDictionary[aa]\n",
        "            posCharge += nAA * ((10 ** self.aa2chargePos[aa])\n",
        "                                / (10 ** self.aa2chargePos[aa] + 10 ** pH))\n",
        "        posCharge += (10 ** self.aaNterm) / (10 ** self.aaNterm + 10 ** pH)  # Include the Nterm to posCharge.\n",
        "\n",
        "        negCharge = 0\n",
        "        for aa in self.aa2chargeNeg:\n",
        "            nAA = self.aaDictionary[aa]\n",
        "            negCharge += nAA * ((10 ** pH)\n",
        "                                / (10 ** self.aa2chargeNeg[aa] + 10 ** pH))\n",
        "        negCharge += (10 ** pH) / (10 ** self.aaCterm + 10 ** pH)\n",
        "\n",
        "        netCharge = posCharge - negCharge\n",
        "\n",
        "        return netCharge\n",
        "\n",
        "    def molarExtinction(self):\n",
        "        \"\"\"Estimates the molar extinction coefficient based on the number\n",
        "        of tyrosines, tryptophans, cysteines and their extinction\n",
        "        coefficient at 280nm which can be found in dictionary(aa2abs280).\n",
        "\n",
        "        Returns:\n",
        "            (float): Molar extinction coefficient.\n",
        "        \"\"\"\n",
        "        tyrosine = self.aaDictionary['Y'] * self.aa2abs280['Y']\n",
        "        tryptophans = self.aaDictionary['W'] * self.aa2abs280['W']\n",
        "        cysteines = self.aaDictionary['C'] * self.aa2abs280['C']\n",
        "        molarEx = tyrosine + tryptophans + cysteines\n",
        "        return molarEx\n",
        "\n",
        "    def massExtinction(self):\n",
        "        \"\"\"Computes mass extinction by dividing molar extinction\n",
        "        by the molecular weight of corresponding protein.\n",
        "\n",
        "        Returns:\n",
        "            (float): Mass extinction value.\n",
        "        \"\"\"\n",
        "        myMW = self.molecularWeight()\n",
        "        return self.molarExtinction() / myMW if myMW else 0.0\n",
        "\n",
        "    def molecularWeight(self):\n",
        "        \"\"\"Calculates the molecular weight of the protein sequence by summing\n",
        "        the weights of the individual Amino acids and excluding the waters\n",
        "        that are released with peptide bond formation.\n",
        "\n",
        "        Returns:\n",
        "            (float): Molecular weight.\n",
        "        \"\"\"\n",
        "        aaWeight = 0\n",
        "        h2o = self.mwH2O * (self.aaCount() - 1)\n",
        "        for aa, count in self.aaDictionary.items():\n",
        "            aaWeight += (count * self.aa2mw[aa])  # Sums the weights of the individual aa's.\n",
        "        return aaWeight - h2o  # Excludes the h2o's released with peptide bond formation.\n",
        "\n",
        "\n",
        "import sys\n",
        "\n",
        "class FastAreader:\n",
        "\n",
        "    def __init__ (self, fname=''):\n",
        "        '''contructor: saves attribute fname '''\n",
        "        self.fname = fname\n",
        "\n",
        "    def doOpen (self):\n",
        "        if self.fname == '':\n",
        "            return sys.stdin\n",
        "        else:\n",
        "            return open(self.fname)\n",
        "\n",
        "    def readFasta (self):\n",
        "\n",
        "        header = ''\n",
        "        sequence = ''\n",
        "\n",
        "        with self.doOpen() as fileH:\n",
        "\n",
        "            header = ''\n",
        "            sequence = ''\n",
        "\n",
        "            # skip to first fasta header\n",
        "            line = fileH.readline()\n",
        "            while not line.startswith('>') :\n",
        "                line = fileH.readline()\n",
        "            header = line[1:].rstrip()\n",
        "\n",
        "            for line in fileH:\n",
        "                if line.startswith ('>'):\n",
        "                    yield header,sequence\n",
        "                    header = line[1:].rstrip()\n",
        "                    sequence = ''\n",
        "                else :\n",
        "                    sequence += ''.join(line.rstrip().split()).upper()\n",
        "\n",
        "\n",
        "        yield header,sequence"
      ],
      "metadata": {
        "id": "RYLoUOSl_Sf8"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def main(myCommandLine=None):\n",
        "    \"\"\"Reads in a fasta file and outputs the ORFs frame, start, stop, and length position on a output file.\"\"\"\n",
        "    if myCommandLine is None:\n",
        "        myCommandLine = CommandLine()\n",
        "        if myCommandLine.args.longestGene:  # If the terminal sees lG flag variable start this part of code.\n",
        "            fastaFile = FastAreader()\n",
        "            for header, sequence in fastaFile.readFasta():\n",
        "                print(header)\n",
        "                orfData = OrfFinder(sequence)\n",
        "                orfData.findOrfs()\n",
        "                orfData.findRevOrfs()\n",
        "                filteredList = filter(lambda orf: orf[3] > myCommandLine.args.minGene, orfData.orfs)  # Filters out the ORFS depending on the minGene arg.\n",
        "                for frame, start, stop, length in sorted(filteredList, key=lambda orf: orf[3], reverse = True):  # Unzips filteredList and sorts the list by length.\n",
        "                    print('{:+d} {:>5d}..{:>5d} {:>5d}'.format(frame, start, stop, length))\n",
        "    else:\n",
        "        myCommandLine = CommandLine(myCommandLine)\n",
        "\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    main()"
      ],
      "metadata": {
        "id": "_hk9EAXO_sDP"
      },
      "execution_count": 6,
      "outputs": []
    }
  ]
}